openedx/core/djangoapps/user_api/completion/tasks.py

```diff
@@ -0,0 +1,181 @@
+    """
+    log.info('Started progress migration from "%s" to "%s" for "%s" course', source, target, course)
+
+    try:
+        course_key = CourseKey.from_string(course)
+    except InvalidKeyError:
+        log.warning('Migration failed. Invalid course key: %s', course)
+        return OUTCOME_COURSE_KEY_INVALID
+
+    try:
+        get_course(course_key)
+    except ValueError:
+        log.warning('Migration failed. Course not found:: %s', course_key)
+        return OUTCOME_COURSE_NOT_FOUND
+
+    try:
+        source = get_user_model().objects.get(email=source)
+    except ObjectDoesNotExist:
+        log.warning('Migration failed. Source user with such email not found: %s', source)
+        return OUTCOME_SOURCE_NOT_FOUND
+
+    try:
+        enrollment = CourseEnrollment.objects.select_for_update().get(user=source, course=course_key)
+    except ObjectDoesNotExist:
+        log.warning(
+            'Migration failed. Source user with email "%s" not enrolled in "%s" course', source.email, course_key
+        )
+        return OUTCOME_SOURCE_NOT_ENROLLED
+
+    try:
+        target = get_user_model().objects.get(email=target)
+    except ObjectDoesNotExist:
+        log.warning('Migration failed. Target user with such email not found: %s', target)
+        return OUTCOME_TARGET_NOT_FOUND
+
+    if CourseEnrollment.objects.filter(user=target, course=course_key).exists():
+        log.warning(
+            'Migration failed. Target user with email "%s" already enrolled in "%s" course', target.email, course_key
+        )
+        return OUTCOME_TARGET_ALREADY_ENROLLED
+
+    # Fetch completions for source user
+    completions = BlockCompletion.user_course_completion_queryset(
+        user=source, course_key=course_key
+    ).select_for_update()
+
+    # Fetch edx-submissions data for source user
+    anonymous_ids = AnonymousUserId.objects.filter(user=source, course_id=course_key).values('anonymous_user_id')
+    submissions = StudentItem.objects.select_for_update().filter(course_id=course_key, student_id__in=anonymous_ids)
+
+    # Fetch StudentModule table data for source user
+    student_states = StudentModule.objects.select_for_update().filter(student=source, course_id=course_key)
+
+    # Actually migrate completions and progress
+    try:
+        with transaction.atomic():
+            # Modify enrollment
+            enrollment.user = target
+            enrollment.save()
+
+            # Migrate completions for user
+            for completion in completions:
+                completion.user = target
+                completion.save()
+
+            # Migrate edx-submissions
+            for submission in submissions:
+                submission.student_id = anonymous_id_for_user(target, course_key)
+                submission.save()
+
+            # Migrate StudentModule
+            for state in student_states:
+                state.student = target
+                state.save()
+
+    except Exception:
```

Comment:
@0x29a Can you be a little bit more transparent about the issue? 
```suggestion
    except Exception as exc:
        log.exception("Unexpected error while migrating user progress: {}".format(str(exc))  # or something similar?
```

Reply:
This will help us debug issues later.

Reply:
@giovannicimolin, `log.exception` itself fetches traceback and throws it into output. Consider this example:
```python
import logging

logger = logging.getLogger(__name__)

def blackbox():
  raise ZeroDivisionError("Hey!")

try:
  blackbox()
except:
  logger.exception("Unexpected error.")

print("The End.")

```

Here is output for this:
```
➜  ~ python3 blackbox.py
Unexpected error.
Traceback (most recent call last):
  File "blackbox.py", line 9, in <module>
    blackbox()
  File "blackbox.py", line 6, in blackbox
    raise ZeroDivisionError("Hey!")
ZeroDivisionError: Hey!
The End.
```

Is that what we want?

Reply:
Ah, nice! Thanks for the heads up on this one.